packages/wallet/src/services/HandlesTracker.ts

```diff
@@ -0,0 +1,89 @@
+import { Assets, HandleInfo } from '../types';
+import { Cardano } from '@cardano-sdk/core';
+import { Logger } from 'ts-log';
+import {
+  NEVER,
+  Observable,
+  combineLatest,
+  distinctUntilChanged,
+  map,
+  mergeMap,
+  of,
+  shareReplay,
+  withLatestFrom
+} from 'rxjs';
+import { deepEquals, isNotNil } from '@cardano-sdk/util';
+import { sameArrayItems, strictEquals } from './util';
+import uniqBy from 'lodash/uniqBy';
+
+export interface HandlesTrackerProps {
+  utxo$: Observable<Cardano.Utxo[]>;
+  assetInfo$: Observable<Assets>;
+  tip$: Observable<Cardano.Tip>;
+  handlePolicyIds: Cardano.PolicyId[];
+  logger: Logger;
+}
+
+const handleInfoEquals = (a: HandleInfo, b: HandleInfo) =>
+  a.assetId === b.assetId &&
+  a.resolvedAt.hash === b.resolvedAt.hash &&
+  deepEquals(a.tokenMetadata, b.tokenMetadata) &&
+  deepEquals(a.nftMetadata, b.nftMetadata);
+
+export const createHandlesTracker = ({ tip$, assetInfo$, handlePolicyIds, logger, utxo$ }: HandlesTrackerProps) =>
+  combineLatest([
+    utxo$.pipe(
+      map((utxo) =>
+        utxo.flatMap(([_, txOut]) =>
+          uniqBy(
+            [...(txOut.value.assets?.keys() || [])]
+              .filter((assetId) => handlePolicyIds.some((policyId) => assetId.startsWith(policyId)))
+              .map((assetId) => ({
+                handleAssetId: assetId,
+                txOut
+              })),
+            ({ handleAssetId }) => handleAssetId
+          )
+        )
+      ),
+      distinctUntilChanged((a, b) => sameArrayItems(a, b, strictEquals)),
+      withLatestFrom(tip$)
+    ),
+    assetInfo$
+  ]).pipe(
+    mergeMap(([[utxo, tip], assets]) => {
+      const handlesWithAssetInfo = utxo
+        .map(({ handleAssetId, txOut }): HandleInfo | null => {
+          const assetInfo = assets.get(handleAssetId);
+          if (!assetInfo) {
+            logger.debug(`Asset info not (yet?) found for ${handleAssetId}`);
+            return null;
+          }
+          return {
+            ...assetInfo,
+            handle: Buffer.from(Cardano.AssetId.getAssetName(handleAssetId), 'hex').toString('utf8'),
+            hasDatum: !!txOut.datum,
+            resolvedAddresses: {
+              cardano: txOut.address
+            },
+            resolvedAt: tip
+          };
+        })
+        .filter(isNotNil);
+      if (utxo.length > 0 && handlesWithAssetInfo.length === 0) {
+        // AssetInfo is still resolving
+        return NEVER;
```

Comment:
I think this will create a leaky/never completing observable every time this branch is reached.
```diff
-        return NEVER;
+        return NEVER;
```

Reply:
Nice catch! :mage: Don't know how to add a test for it but fixed it in ed43d33504bc5591954a035b20abf6f1a396d61e